smolyay/basis.py

```diff
@@ -403,29 +482,393 @@ def _derivative(self, x):
+        points = self.domain[0] + (self.domain[1] - self.domain[0]) * (
+            (points - old_domain[0]) / (old_domain[1] - old_domain[0])
+        )
+        return numpy.clip(points, self.domain[0], self.domain[1])
+
+    def __call__(self, X, X_domain):
+        """Evaluate all the basis functions in the set
+
+        Calls all the basis function(s) at index and evaluates at X.
+
+        Parameters
+        ----------
+        X : array-like
+            the points to evaluate.
+
+        X_domain : numpy array of shape (2,)
+            the lower and upper bounds of X.
+
+        Returns
+        -------
+        scalar or ndarray
+            the values of the basis functions.
+        """
+
+        new_X = self._scale_to_domain(numpy.asarray(X), X_domain)
+        if any(bf._is_complex for bf in self):
+            y = numpy.zeros([len(self)] + list(new_X.shape), dtype="complex_")
+        else:
+            y = numpy.zeros([len(self)] + list(new_X.shape))
+        for i in range(len(self)):
+            y[i, :] = self[i](new_X)
+        return y
+
+    def derivative(self, X, X_domain, n=1):
+        """Evaluate all the derivative of basis functions in the set
+
+        Calls the derivative for all the basis function(s) and
+        evaluates at X.
+
+        Parameters
+        ----------
+        X : array-like
+            the points to evaluate.
+
+        X_domain : numpy array of shape (2,)
+            the lower and upper bounds of X.
+
+        n : int, optional
+            order of derivative. Default is 1.
+
+        Returns
+        -------
+        scalar or ndarray
+            the values of the basis functions.
+        """
+        new_X = self._scale_to_domain(numpy.asarray(X), X_domain)
+        if any(bf._is_complex for bf in self):
+            y = numpy.zeros([len(self)] + list(new_X.shape), dtype="complex_")
+        else:
+            y = numpy.zeros([len(self)] + list(new_X.shape))
+        for i in range(len(self)):
+            y[i, :] = self[i].derivative(new_X, n)
+        y *= ((self.domain[1] - self.domain[0]) / (X_domain[1] - X_domain[0])) ** n
+        return y
+
+
+class CustomBasisFunctionSet(BasisFunctionSet):
+    """Set of basis functions
 
     Parameters
     ----------
     basis_functions : list
         Basis functions in set.
 
+    Raises
+    ------
+    IndexError
+        Must have at least one basis function
+    ValueError
+        Basis functions must have the same domain.
     """
 
     def __init__(self, basis_functions):
+        if len(basis_functions) == 0:
+            raise IndexError("Must have at least one basis function.")
+        domain = basis_functions[0].domain
+        if any(not numpy.array_equal(domain, b.domain) for b in basis_functions[1:]):
+            raise ValueError("Basis functions must have the same domain.")
         self._basis_functions = basis_functions
 
+
+class MutableBasisFunctionSet(BasisFunctionSet):
```

Comment:
I don't think this class will be needed either. See comment below on structuring the constructors, and then `num_terms` is redundant to your `__len__` method. If we want to keep `num_terms`, it can go in the `BasisFunctionSet` class.

Reply:
I'm fine with generating `basis_functions` in the constructor. The only issue is that `num_terms` is currently mutable, which in the context of `BasisFunctionSet` doesn't make sense since it takes a list of basis functions, not the number wanted.

We could have a class parameter that designates if setting `num_terms` should be allowed, which is False for the base class but True for others. Or perhaps classes like `ChebyshevFirstKindBasisFunctionSet` have a public method that takes a `num_terms` and generate the basis functions from that. For the nested sets this takes `num_levels` instead. This public method replaces the need for setters.

Reply:
`num_terms` serves basically the same purpose as `num_points` in the random point set. Both return an equivalent value to `__len__`, but they exist to make that property mutable.

Reply:
It won't be mutable for a basis function set, but sure, you can have this as a read-only property if people prefer it to len.

Reply:
Sorry, I didn't see the comment above: `num_terms` should not be mutable in the base class. Deriving classes *could* implement a setter if they had a reason for it (it would slice/extend the the basis function list), but for now, let's not worry about being able to do that.

Reply:
Okay. Mutability was the main issue and was the reason I didn't generate everything in the constructor in the first place.